### ISSUE TITLE
hasMany aync relationship does not load, throws exception

### ISSUE THREAD

**Issue by aortbals** [emberjs/data#2662]
This regression appears in Ember Data 1.0.0-beta.14. It seems that `flushCanonical` is triggering an exception in Ember when it tries to [destroy child views that don't exist](https://github.com/emberjs/ember.js/blob/master/packages/ember-views/lib/views/collection_view.js#L326)?

Regression in Ember Data 1.0.0.-beta.14: http://jsbin.com/pagotu/1/edit?html,js,console,output
Works correctly in Ember Data 1.0.0.-beta.12: http://jsbin.com/vobevevapo/3/edit?html,js,console,output
Also appears in Ember Canary (actually prints the attribute for the hasMany record, still throws exception): http://jsbin.com/pagotu/3/edit?html,js,console,output
It also gets thrown on a basic `hasMany` <-> `belongsTo` relationship with `links` and `async: true`: http://jsbin.com/pagotu/5/edit?html,js,console,output

```
"TypeError: Cannot read property 'destroy' of undefined
    at ContainerView.extend.arrayWillChange (http://builds.emberjs.com/release/ember.js:40513:20)
    at applyStr (http://builds.emberjs.com/release/ember.js:19678:29)
    at sendEvent (http://builds.emberjs.com/release/ember.js:14114:13)
    at __exports__.default.Mixin.create.arrayContentWillChange (http://builds.emberjs.com/release/ember.js:30442:9)
    at EmberObject.extend.arrangedContentArrayWillChange (http://builds.emberjs.com/release/ember.js:34309:14)
    at applyStr (http://builds.emberjs.com/release/ember.js:19678:29)
    at sendEvent (http://builds.emberjs.com/release/ember.js:14114:13)
    at __exports__.default.Mixin.create.arrayContentWillChange (http://builds.emberjs.com/release/ember.js:30442:9)
    at Ember.Object.extend.flushCanonical (http://builds.emberjs.com/canary/ember-data.js:4784:14)
    at ember$data$lib$system$relationships$state$has_many$$ManyRelationship.flushCanonical (http://builds.emberjs.com/canary/ember-data.js:6508:22)"
```

**Comment by igorT** [comment]
Thanks for the bug report, it appears to be a real bug. Investigating

**Comment by MarSik** [comment]
Is there any workaround I can use (ideally in ember cli environment) before a proper fix is released? This is breaking my project (fortunately no customers depend on it) but I can't easily downgrade, because beta-14 fixed other array and polymorphic related issues.

**Comment by teddyzeenny** [comment]
@MarSik two possible ways you can work around it now, any of these two should work:

1. Make sure your ManyArray is fully loaded before rendering a template:
```js
export default Ember.Route.extend({
  model() {
    return this.store.find('post', 1);
  },
  afterModel(model) {
    return model.get('comments');
  }
})
```

2. Render a clone array of the ManyArray
```js
  export default Ember.Controller.extend({
    comments: Ember.computed.map('model.comments', item => item) 
  });
```
```handlebars
  {{#each comment in comments}}
  {{/each}}
```

**Comment by wecc** [comment]
I did a lot of research on this issue and I think there's something fishy going on scheduling calls to `Relationship.flushCanonical` (using Backburner, `syncRelationships` queue), it doesn't get scheduled but called straight away which makes (in the case of your failing JSBin) `ManyArray.flushCanonical` getting called six times for the same relationship. And since `canonicalState` and `currentState` on the first call is empty, when it propagates down to ContainerView that tries to remove something that doesn't exist, it explodes.

I'm sorry if this makes no sense at all, but I would love to dig deeper into this and maybe get a bit of wisdom from the core team on how to proceed.

**Comment by sly7-7** [comment]
It looks like this issue is a pretty tough one, but also a pretty common one... I encountered this myself when trying to upgrade ED on the app I'm now working on.
Do we have any idea on how to resolve this ? I really think this should be fixed before the next beta.
cc/ @igorT @bmac

**Comment by bmac** [comment]
@sly7-7 I just tested this jsbin with the changes in https://github.com/emberjs/data/pull/2723 and the seem to fix the issue.

**Comment by sly7-7** [comment]
That would be great :+1: What do you need to merge it ? @igorT's review ?

**Comment by bmac** [comment]
@sly7-7 I just need someone to review it. Don't feel comfortable merging my own stuff without someone else's +1. Your eyes would be great if you have time.

**Comment by bmac** [comment]
should be fixed by #2723
